rust/src/cliwrap/yumdnf.rs

```diff
@@ -0,0 +1,152 @@
+    Upgrade,
+    /// Start an upgrade of the operating system
+    Update,
+    /// Display information about system state
+    Status,
+    /// Perform a search of packages.
+    Search {
+        /// Search terms
+        terms: Vec<String>,
+    },
+    /// Will return an error suggesting other approaches.
+    Install {
+        /// Set of packages to install
+        packages: Vec<String>,
+    },
+    /// Layer RPM packages as operating system extensions
+    InstallExtension {
+        /// Set of packages to install
+        packages: Vec<String>,
+    },
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum RunDisposition {
+    HelpOrVersionDisplayed,
+    ExecRpmOstree(Vec<String>),
+    UseSomethingElse(&'static str),
+    Unhandled,
+}
+
+fn disposition(argv: &[&str]) -> Result<RunDisposition> {
+    let opt = match Opt::from_iter_safe(std::iter::once(&"yum").chain(argv.iter())) {
+        Ok(v) => v,
+        Err(e)
+            if e.kind == clap::ErrorKind::VersionDisplayed
+                || e.kind == clap::ErrorKind::HelpDisplayed =>
+        {
+            return Ok(RunDisposition::HelpOrVersionDisplayed)
+        }
+        Err(_) => {
+            return Ok(RunDisposition::Unhandled);
+        }
+    };
+
+    let disp = match opt {
+        Opt::Upgrade | Opt::Update => RunDisposition::ExecRpmOstree(vec!["upgrade".into()]),
+        Opt::Status => RunDisposition::ExecRpmOstree(vec!["status".into()]),
+        Opt::Install { packages: _ } => {
+            // TODO analyze packages to find e.g. `gcc` (not ok, use `toolbox`) versus `libvirt` (ok)
+            RunDisposition::UseSomethingElse(indoc! {r##"
+This system is rpm-ostree based.  Before installing packages to
+the host root filesystem, consider other options:
+
+ - `toolbox`: For command-line development and debugging tools in a privileged container
+ - `podman`: General purpose containers (run a webserver, e.g.)
+ - `flatpak`: For desktop (GUI) applications
+ - `yum install-extension`: The same thing as `rpm-ostree install --apply-live`.  Treat packages as OS extensions.
+"##})
+        }
+        Opt::Search { .. } => RunDisposition::UseSomethingElse(indoc! { r##"
+This system is rpm-ostree based.  Package search is not yet implemented.
+For now, it's recommended to use e.g. `toolbox` and `dnf search` inside there.
+            "##}),
+        Opt::InstallExtension { packages } => RunDisposition::ExecRpmOstree(
+            ["install", "--apply-live"]
+                .iter()
+                .map(|&s| s)
+                .chain(packages.iter().map(|s| s.as_str()))
+                .map(|s| s.to_string())
+                .collect(),
+        ),
+    };
+
+    Ok(disp)
+}
+
+/// Primary entrypoint to running our wrapped `yum`/`dnf` handling.
+pub(crate) fn main(argv: &[&str]) -> Result<()> {
+    match disposition(argv)? {
+        RunDisposition::HelpOrVersionDisplayed => Ok(()),
+        RunDisposition::ExecRpmOstree(args) => {
```

Comment:
I think here is where we would print a redirection message with the underlying rpm-ostree command being used?

Reply:
My instinct here is we're explicitly not trying to always force people to rewrite their CLI.  Being called as `yum` should IMO feel as equally first class as being invoked as `rpm-ostree` in some cases.  So for now, this PR just goes with the current proposed text that prints `Note: This system is image (rpm-ostree) based.`.

Reply:
Yeah, I think that makes sense. (I originally mentioned `dnf` printing a redirection warning as an argument for doing it here too, and while I'm pretty sure it did do that at some point, it looks like it no longer does.)